### PR TITLE
Major refactor of functions deploy

### PR DESCRIPTION
**Pull request by joehan**
### Description
Refactor of most of the functions deploy code path to use Typescript & more modern coding style, and to take advantage of new GCF features that will reduce the number of builds needed for a deployment.

All the code changes here have been reviewed in separate PRs - however, it still may be worth a quick scan through.

### Scenarios Tested
A whole lot of them - I ran a large number of deploys of different combinations of creates/updates/delete on different types of functions, and we ran a bug bash. I also tested some specific situations that were likely to cause issues - new projects, projects without billing enabled, projects without the appropriate APIs enabled, etc.

### CODE REVIEW

**Comment by inlined** [comment]
Is this just a merge branch for changes that have already been reviewed up to now? If so I can robo-LGTM

**Comment by joehan** [comment]
@inlined Yes it is - everything in here has already been reviewed.

**Comment by inlined** [comment]
Then I hereby robo-stamp

**Review by inlined — APPROVED** [review]

File: src/deploy/functions/deploymentPlanner.ts [hunk 1]
```diff
@@ -0,0 +1,181 @@
+import * as deploymentTool from "../../deploymentTool";
+import { functionMatchesAnyGroup, getTopicName } from "../../functionsDeployHelper";
+import { checkForInvalidChangeOfTrigger } from "./validate";
+
+// TODO: Better name for this?
+// It's really a CloudFuntion, not just a trigger,
+// but CloudFunction is a different exported type from firebase-functions
+export interface CloudFunctionTrigger {
```
Comment (IchordeDionysos) [inline src/deploy/functions/deploymentPlanner.ts:8]: @joehan how about `CloudFunctionConfig`?

**Review by IchordeDionysos — COMMENTED** [review]
@joehan I am really interested in what you've tested.

Did you test to deploy more than a couple of functions like 10-25 functions at once and how did it perform in terms of speed of deployment and reliability of the deployments?

In the past, we've seen deployments of more than a handful of functions to be less reliable and tend to fail more often.
So, I am interested if this change would improve this?

**Comment by mbleigh** [comment]
@IchordeDionysos this change should substantially help with reliability in the scenario you described. I'd encourage you to update to the newest CLI version (v9.4.0) and try it out!

**Comment by joehan** [comment]
@IchordeDionysos We tested deploys of 100+ functions at once, and saw substantially better performance and reliability than with previous versions. Different functions may have different results, depending on things like how big your dependencies are, etc, but I suspect this should help out in the scenario you described!

**Comment by rhodgkins** [comment]
Not sure if this is the right place (have also commented in #2060) but still getting the quota errors 40-60% of the time when deploying 80 functions.

**Pull request merged**